qiskit/passmanager/base_pass.py

```diff
@@ -60,7 +60,24 @@ def __init__(self):
     def __hash__(self):
         return self._hash
 
+    def __iter__(self):
+        # To avoid normalization overhead.
+        # A pass must be scheduled in iterable format in the pass manager,
+        # which causes serious normalization overhead, since passes can be nested.
+        # This iterator allows the pass manager to treat BasePass and FlowController equally.
+        yield self
+
+    def __len__(self):
+        return 1
+
     def __eq__(self, other):
+        try:
+            if len(other) == 1:
+                # Reference might be List[BasePass] which was conventional representation
+                # of a single base pass in the pass manager.
+                other = next(iter(other))
+        except TypeError:
+            return False
         return hash(self) == hash(other)
```

Comment:
Yes, it's actually used. Pass runner (currently in flow controller) manages already run passes with a set instance, and it prevents execution of duplicated passes by design. This equivalence is used for the set object. For example,

https://github.com/Qiskit/qiskit-terra/blob/28113b6a9fc1fd31bc211a73fe7a6908fffa1114/test/python/transpiler/test_pass_scheduler.py#L798-L819

The actual log without equivalency will be

```
run transformation pass PassA_TP_NR_NP
Pass: PassA_TP_NR_NP - 0.46515 (ms)   <-- required
run transformation pass PassC_TP_RA_PA
Pass: PassC_TP_RA_PA - 0.05412 (ms)
run transformation pass PassA_TP_NR_NP
Pass: PassA_TP_NR_NP - 0.04673 (ms)   <-- required
run transformation pass PassB_TP_RA_PA
Pass: PassB_TP_RA_PA - 0.04196 (ms)
...
```

Since both `PassC_TP_RA_PA` and `PassB_TP_RA_PA` require `PassA_TP_NR_NP`, which is independently instantiated and added to `.requires`.  Dropping this violates ~30 unit tests, and may impact user's code by repeatedly executing the requires passes.

Strictly speaking this implementation is not really robust. This logic uses Python repr as a easy hash, however every Qiskit object doesn't properly implement the repr dunder (e.g. `Target`). Developer should have implemented the equality dunder for every pass, but it's too late now.